### DISCUSSION TITLE
Have syntactic sugar for 'class' (+ 'struct') and 'unmanaged' extension methods

### DISCUSSION THREAD

**Discussion by john-h-k** [dotnet/csharplang#2086] [Language Ideas]
Restricting extension methods too a base, interface, or struct class can be done simply by (obviously) defining the `this` parameter as the proper base/interface or `System.ValueType` for structs. It is also possible to constrain extension methods too `class` and `unmanaged` via generics, as such:
```
public void Function<T>(this T obj) where T : unmanaged
```
I suggest that these function signatures below should be equivalent to their generic counterparts

```
Function(this struct obj);
equivalent to
Function<T>(this T obj) where T : struct;
and
Function(this System.ValueType obj);
```
For reference types
```
Function(this class obj);
equivalent to
Function<T>(this T obj) where T : class;
```
For unmanaged types, the same, just with the `unmanaged` keyword.

## Why?
* A cleaner syntax that loses no information
* Easier to search for (simply searching for `this [keyword]` rather than a generic type constraint with regex to ignore the name and other parameters etc). 
* Syntax fits alongside generic type constraint syntax
* Use of `struct`, while equal in length / complexity to `ValueType` signature, is arguably clearer 
* Wouldn't require any CLR changes
* Limits the instance type information of an extension method to keywords in the first method, rather than spread across generic type constraints which potentially constrain other generic types

**Comment by HaloFour** [comment]
> A cleaner syntax that loses no information

It loses the information that `Function` is a generic member with a generic arity of 1, which is critically important as that is a part of the member name and signature.  This proposal doesn't suggest how to reconcile that with multiple parameters or even multiple generic type parameters.
